// @http2/server/src/helpers/normaliseManifest.js
// https://tools.ietf.org/html/rfc7540#section-5.3.5
// All streams are initially assigned a non-exclusive dependency on
// stream 0x0.  Pushed streams (Section 8.2) initially depend on their
// associated stream.  In both cases, streams are assigned a default
// weight of 16.
const DEFAULT_PRIORITY = 16

function isGlob (value) {
  return typeof value === 'string' && !value.startsWith('https://')
}

function validRule ({uri, glob, push}) {
  return push.length && ((uri && uri.length) || (glob && glob.length))
}

function validPush ({uri, glob}) {
  return (uri && uri.length) || (glob && glob.length)
}

module.exports.normaliseManifest = (manifest) => {
  return (manifest || []).map((rule) => {
    for (const key of ['uri', 'glob', 'push']) {
      if (typeof rule[key] === 'string') rule[key] = [rule[key]]
    }
    if (Array.isArray(rule.push) && rule.push.every(isGlob)) {
      rule.push = [{
        priority: DEFAULT_PRIORITY,
        glob: Array.from(rule.push)
      }]
    } else {
      rule.push = (rule.push || []).map((dep) => {
        if (typeof dep === 'string') {
          if (dep.startsWith('https://')) dep = {uri: [dep]}
          else dep = {glob: [dep]}
        }
        if (typeof dep.uri === 'string') dep.uri = [dep.uri]
        if (typeof dep.glob === 'string') dep.glob = [dep.glob]
        if (typeof dep.priority !== 'number') dep.priority = DEFAULT_PRIORITY
        return dep
      }).filter(validPush)
    }
    return rule
  }).filter(validRule)
}
