// @http2/configuration/schema/host.js
module.exports.hostSchema = {
  id: '/Host',
  type: 'object',
  additionalProperties: false,
  properties: {
    domain: {
      anyOf: [
        {type: 'string', format: 'hostname'},
        {type: 'string', format: 'ipv4'},
        {type: 'string', format: 'ipv6'}
      ]
    },
    root: {type: 'string'},
    fallback: {type: 'string'},
    cacheControl: {
      type: 'object',
      additionalProperties: false,
      properties: {
        immutable: {
          type: 'array',
          uniqueItems: true,
          items: {type: 'string', format: 'extglob'}
        }
      }
    },
    directories: {
      type: 'object',
      additionalProperties: false,
      properties: {
        trailingSlash: {
          enum: ['always', 'never']
        }
      }
    },
    accessControl: {
      type: 'object',
      additionalProperties: false,
      properties: {
        allowOrigin: {
          anyOf: [
            {type: 'string', enum: ['*']},
            {type: 'string', format: 'uri'}
          ]
        }
      }
    },
    serviceWorker: {
      type: 'object',
      additionalProperties: false,
      properties: {
        allowed: {type: 'string'}
      }
    },
    manifest: {'$ref': '/Manifest'}
  }
}
