// @http2/configuration/schema/manifest.js
module.exports.manifestSchema = {
  id: '/Manifest',
  type: 'array',
  definitions: {
    uri: {
      anyOf: [
        {type: 'string', format: 'uri-template'},
        {
          type: 'array',
          minItems: 1,
          uniqueItems: true,
          items: {type: 'string', format: 'uri-template'}
        }
      ]
    },
    glob: {
      anyOf: [
        {type: 'string', format: 'extglob'},
        {
          type: 'array',
          minItems: 1,
          uniqueItems: true,
          items: {type: 'string', format: 'extglob'}
        }
      ]
    }
  },
  additionalProperties: false,
  items: {
    type: 'object',
    additionalProperties: false,
    required: ['push'],
    oneOf: [
      {required: ['uri']},
      {required: ['glob']}
    ],
    properties: {
      uri: {'$ref': '#/definitions/uri'},
      glob: {'$ref': '#/definitions/glob'},
      push: {
        anyOf: [
          {
            type: 'string',
            anyOf: [
              {format: 'uri-template'},
              {format: 'extglob'}
            ]
          },
          {
            type: 'array',
            minItems: 1,
            uniqueItems: true,
            items: {
              anyOf: [
                {
                  type: 'string',
                  anyOf: [
                    {format: 'uri-template'},
                    {format: 'extglob'}
                  ]
                },
                {
                  type: 'object',
                  properties: {
                    priority: {type: 'integer', minimum: 1, maximum: 256},
                    uri: {'$ref': '#/definitions/uri'},
                    glob: {'$ref': '#/definitions/glob'}
                  }
                }
              ]
            }
          }
        ]
      }
    }
  }
}
